src/codec/tremor.rs

```diff
@@ -0,0 +1,423 @@
+// Copyright 2020-2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The Tremor codec is meant for high performance interfacing between tremor instances
+
+use std::{
+    fmt,
+    io::{Cursor, Write},
+};
+
+use super::prelude::*;
+use beef::Cow;
+use byteorder::{ByteOrder, NetworkEndian, ReadBytesExt, WriteBytesExt};
+use halfbrown::HashMap;
+use simd_json::StaticNode;
+
+#[derive(Clone)]
+pub struct Tremor {}
+
+type DefaultByteOrder = NetworkEndian;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Error {
+    /// Invalid data
+    InvalidData,
+    /// Invalid version
+    InvalidVersion,
+}
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidData => write!(f, "Invalid data"),
+            Self::InvalidVersion => write!(f, "Invalid version"),
+        }
+    }
+}
+impl std::error::Error for Error {}
+
+impl Tremor {
+    const NULL: u8 = 0;
+    const BOOL_TRUE: u8 = 1;
+    const BOOL_FALSE: u8 = 2;
+    const I64: u8 = 3;
+    const U64: u8 = 4;
+    const F64: u8 = 5;
+    const STRING: u8 = 6;
+    const BYTES: u8 = 7;
+    const ARRAY: u8 = 8;
+    const OBJECT: u8 = 9;
+    const VERSION: u32 = 0;
```

Comment:
you can totally make this a `u8`. We likely won't make it past version 255 anytime...

Reply:
that's what they said with the Y2K problem :P lets pick the middle and make it 16?

Reply:
your call :)